develop/reference_sql.rst

```diff
@@ -12,7 +12,46 @@ In the following sections, we discuss the different types of queries you can run
 Aggregate Functions
 -------------------
 
-Citus supports and parallelizes most aggregate functions supported by PostgreSQL. Citus's query planner transforms the aggregate into its commutative and associative form so it can be parallelized. In this process, the workers run an aggregation query on the shards and the coordinator then combines the results from the workers to produce the final output.
+Citus supports and parallelizes most aggregate functions supported by
+PostgreSQL. Citus's query planner transforms the aggregate into its commutative
+and associative form so it can be parallelized. In this process, the workers
+run an aggregation query on the shards and the coordinator then combines the
+results from the workers to produce the final output.
+
+Aggregate execution works in two ways:
+
+1. When the aggregate is grouped by a table's distribution column, and there
+   are no aggregates on distinct values of non-distribution columns, Citus can
+   push down execution of the entire query to each worker. All PostgreSQL
+   aggregates and custom aggregates are supported in this situation, and execute
+   in parallel on the worker nodes.
+2. Otherwise, if grouping on a non-distribution column, or aggregating on a
```

Comment:
This is an incomplete set of when the coordinator has to get involved. It's also done for aggregates such as `array_agg(col ORDER BY ...)`, & aggregates we either don't explicitly support or don't support with COMBINEFUNC (not exactly 1 parameter, not pseudotype `STYPE`)

Reply:
Is `array_agg` the culprit in your example, or the fact that it's operating on an ordered column?

Can you explain more about the reasoning behind COMBINEFUNC restrictions?

I'm looking for the general idea behind these cases. I can present that idea, and then say "for example ... [cases]"

Reply:
The fact that it's operating on an ordered column is what prevents distribution

https://github.com/citusdata/citus/pull/2921 is the PR for COMBINEFUNC. Single argument is used because we have helper functions which have to approximate the signature of the aggregate. Supporting multiple arguments requires implementing generation of the helper functions at runtime, which we decided to not do. Rejecting pseudotype transition state is done because that simplifies serialize/deserialize. We can't really be sure that an `INTERNAL` serialized format is safe to send to a different db (it may contain OIDs etc)

Reply:
I updated scenario two. Is my description accurate enough at a high level?

Reply:
Second scenario seems well covered. Skips over COMBINEFUNC, but that's somewhat being addressed with "Citus supports and parallelizes most aggregate functions supported by PostgreSQL"

I think the mix up in this is that we actually have 3 ways of executing aggregates, 2 of which are new. The two new ways are what you've outlined. The third is partial aggregation, which is what the paragraph before this list is getting at